api/src/migrations/convert-translation-messages.js

```diff
@@ -0,0 +1,82 @@
+  db = require('../db-nano'),
+  properties = require('properties');
+
+module.exports = {
+  name: 'convert-translation-messages',
+  created: new Date(2018, 11, 8),
+  run: promisify((callback) => {
+	  db.request({
+      db: db.settings.db,
+      method: 'GET',
+      path: '_design/medic',
+    },
+    (err, result) => {
+      if (err) {
+        return callback(err);
+      }
+
+      let translationAttachmentKeys = Object.keys(result._attachments).filter(k => k.includes('translations'));
+      translationAttachmentKeys.forEach((translationAttachmentKey) => {
+        let translationMessageKey = translationAttachmentKey.match(/translations\/(.+).properties/)[1];
+        let translationMessageKeys = [translationMessageKey, translationMessageKey + '-backup'];
+        translationMessageKeys.forEach((translationMessageKey) => {
+          db.request({
+            db: db.settings.db,
+            method: 'GET',
+            path: translationMessageKey,
+          },
+          (err, translationMessageValues) => {
+            if (_.has(translationMessageValues, 'values')) {
+              db.request({
+                db: db.settings.db,
+                method: 'GET',
+                path: '_design/medic/' + translationAttachmentKey,
+              },
+              (err, translationAttachmentValue) => {
+                if (err) {
+                  return callback(err);
+                }
+
+                properties.parse(translationAttachmentValue.toString('utf8'), (err, translationAttachmentValues) => {
+                  if (err) {
+                    return callback(err);
+                  }
+
+                  let defaultTranslations = {};
+                  let customTranslations = {};
+                  Object.keys(translationMessageValues.values).forEach((translationMessageKey) => {
+                    if (_.has(translationAttachmentValues, translationMessageKey)) {
+                      defaultTranslations[translationMessageKey] = translationMessageValues.values[translationMessageKey];
+                    } else {
+                      customTranslations[translationMessageKey] = translationMessageValues.values[translationMessageKey];
+                    }
+                  });
+
+                  delete translationMessageValues.values;
+                  translationMessageValues.custom = customTranslations;
+                  translationMessageValues.default = defaultTranslations;
+
+                  db.request({
+                    db: db.settings.db,
+                    method: 'PUT',
+                    path: translationMessageKey,
+                    body: translationMessageValues
+                  },
+                  (err) => {
+                    if (err) {
+                      return callback(err);
+                    }
+                  });
+                });
+              });
+            }
+          });
+        });
+      });
+
+      callback();
+    });
+  })
+};
```

Comment:
We need to add an integration test to ensure this is working as expected. Have a look in `api/tests/integration/migrations` and use that format.

If you'd like you can also write a unit test but integration tests are more helpful here.

This is particularly useful if you want to test the current implementation _before_ you do any refactoring to ensure you don't accidentally break anything!

Reply:
👍